src/meta-srv/src/handler.rs

```diff
@@ -103,13 +114,219 @@ impl HeartbeatHandlerGroup {
+                break;
+            }
+
+            let role = req.header.as_ref().and_then(|h| Role::from_i32(h.role));
+            if h.is_acceptable(role) {
+                h.handle(&req, &mut ctx, &mut acc).await?;
+            }
         }
         let header = std::mem::take(&mut acc.header);
         let res = HeartbeatResponse {
             header,
-            payload: acc.into_payload(),
+            mailbox_messages: acc.into_mailbox_messages(),
         };
         Ok(res)
     }
 }
+
+struct Tx(oneshot::Sender<Result<MailboxMessage>>);
+struct Rx(oneshot::Receiver<Result<MailboxMessage>>);
+
+pub struct HeartbeatMailbox {
+    pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+    sequence: Sequence,
+    senders: DashMap<MessageId, Tx>,
+    recervers: DashMap<MessageId, Rx>,
+    timeouts: DashMap<MessageId, Duration>,
+}
+
+impl HeartbeatMailbox {
+    pub fn with_timeout(
+        pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+        sequence: Sequence,
+        check_interval_millis: u64,
+    ) -> MailboxRef {
+        let ins = Arc::new(Self {
+            pushers,
+            sequence,
+            senders: DashMap::default(),
+            recervers: DashMap::default(),
+            timeouts: DashMap::default(),
+        });
+
+        if check_interval_millis > 0 {
+            let timeout_checker = ins.clone();
+            common_runtime::spawn_bg(async move {
+                timeout_checker
+                    .check_timeout_bg(check_interval_millis)
+                    .await;
+            });
+        }
+
+        ins
+    }
+
+    async fn check_timeout_bg(&self, interval_millis: u64) {
+        let mut interval = tokio::time::interval(Duration::from_millis(interval_millis));
+
+        loop {
+            interval.tick().await;
+
+            let now = Duration::from_millis(common_time::util::current_time_millis() as u64);
+            let timeout_ids = self
+                .timeouts
+                .iter()
+                .filter_map(|entry| {
+                    let (id, deadline) = entry.pair();
+                    if deadline < &now {
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            for id in timeout_ids {
+                let _ = self
+                    .on_recv(id, Err(error::MailboxTimeoutSnafu { id }.build()))
+                    .await;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Mailbox for HeartbeatMailbox {
+    async fn send(&self, ch: &Channel, mut msg: MailboxMessage) -> Result<MessageId> {
+        let message_id = self.sequence.next().await?;
+
+        msg.id = message_id;
+        let res = HeartbeatResponse {
+            header: None,
+            mailbox_messages: vec![msg],
+        };
+
+        let pusher_id = match ch {
+            Channel::Datanode(id) => format!("0-{}", id),
```

Comment:
Seems "0" and "1" are `Role`'s enum index defined in proto. Better extract some helper or util methods to make it clear. Also to "src/meta-srv/src/service/heartbeat.rs" line 53